### PR TITLE
Refactor to increase readability

### LINKED ISSUE #271: Add MARK comments to large files / Split files using extensions
Interested in opinions on this - I think it will greatly improve readability and ease of navigating through the source files.

### PR DESCRIPTION
**Pull request by DJRHails**
Fixes #271 

# Implementation Overview
Based on discussions with @bearbin, certain aspects of the flint codebase are getting difficult to grok as a newcomer. We want to encourage contributions and as such want a highly readable compiler. As part of this I've also started to document the compiler itself.

See new [compiler guide](https://github.com/DJRHails/flint/blob/pass_refactor/docs/compiler_guide.md) to see changes.

# Checklist

- [ ] Remove all TODOs, debug prints, whitespace changes
- [ ] Add `review wanted` label, remove "WIP" in PR title

### CODE REVIEW

File: Sources/Diagnostic/DiagnosticsVerifier.swift [hunk 1]
```diff
@@ -8,16 +8,18 @@
...
 /// Verifies the diagnostics emitted by a program matches exactly what was expected.
 /// The expected diagnostics are specified inline in the source file.
-struct DiagnosticsVerifier {
+public struct DiagnosticsVerifier {
   private let diagnosticRegex = try! NSRegularExpression(pattern: "//\\s*expected-(error|note|warning)\\s*\\s+\\{\\{(.*)\\}\\}")
   private let diagnosticLineRegex = try! NSRegularExpression(pattern: "//\\s*expected-(error|note|warning)\\s*@(\\d+)\\s+\\{\\{(.*)\\}\\}")
   private let diagnosticOffsetRegex = try! NSRegularExpression(pattern: "//\\s*expected-(error|note|warning)\\s*@(-|\\+)(\\d+)\\s+\\{\\{(.*)\\}\\}")
 
-  func verify(producedDiagnostics: [Diagnostic], compilationContext: CompilationContext) -> Bool {
+  public init(){}
```
Comment (bearbin) [inline Sources/Diagnostic/DiagnosticsVerifier.swift]: Spacing

File: Sources/AST/ASTPass/ASTPass.swift [hunk 2]
```diff
@@ -1,354 +1,341 @@
...
+  func process(literalToken: Token, passContext: ASTPassContext) -> ASTPassResult<Token>
+  func process(attribute: Attribute, passContext: ASTPassContext) -> ASTPassResult<Attribute>
+  func process(parameter: Parameter, passContext: ASTPassContext) -> ASTPassResult<Parameter>
+  func process(typeAnnotation: TypeAnnotation, passContext: ASTPassContext) -> ASTPassResult<TypeAnnotation>
+  func process(identifier: Identifier, passContext: ASTPassContext) -> ASTPassResult<Identifier>
+  func process(type: Type, passContext: ASTPassContext) -> ASTPassResult<Type>
+  func process(callerCapability: CallerCapability, passContext: ASTPassContext) -> ASTPassResult<CallerCapability>
+  func process(typeState: TypeState, passContext: ASTPassContext) -> ASTPassResult<TypeState>
+
+  // MARK: -
```
[the remark this answers is not in the document]
Comment (DJRHails) [reply Sources/AST/ASTPass/ASTPass.swift]: The dash mark is a section separator.

**Comment by DJRHails** [comment]
I do agree with you that + is unusually. The swift codebases I've looked at however use it pretty consistently in the same way we've done it.  So we could do the _ but we'd be going away from the Swift common code style.

File: Sources/AST/ASTPass/ASTPass.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (bearbin) [reply Sources/AST/ASTPass/ASTPass.swift]: OK, makes sense.

**Comment by bearbin** [comment]
I don't mind the `+` that much, if there is a good reason (language consistency is good) then for all means keep them.

**Comment by SusanEisenbach** [comment]
Good to see better modularisation.

**Pull request merged**